src/sinks/aws_cloudwatch_metrics.rs

```diff
@@ -166,51 +170,58 @@ impl CloudWatchMetricsSvc {
-                            dimensions,
-                            ..Default::default()
-                        }),
-                        MetricValue::Set { values } => Some(MetricDatum {
-                            metric_name,
-                            value: Some(values.len() as f64),
-                            timestamp,
-                            dimensions,
-                            ..Default::default()
-                        }),
-                        _ => None,
-                    },
-                    MetricKind::Absolute => match event.data.value {
-                        MetricValue::Gauge { value } => Some(MetricDatum {
-                            metric_name,
-                            value: Some(value),
-                            timestamp,
-                            dimensions,
-                            ..Default::default()
-                        }),
-                        _ => None,
-                    },
+                // AwsCloudwatchMetricNormalize converts these to the right MetricKind
+                match event.data.value {
+                    MetricValue::Counter { value } => Some(MetricDatum {
+                        metric_name,
+                        value: Some(value),
+                        timestamp,
+                        dimensions,
+                        ..Default::default()
+                    }),
+                    MetricValue::Distribution {
+                        samples,
+                        statistic: _,
+                    } => Some(MetricDatum {
+                        metric_name,
+                        values: Some(samples.iter().map(|s| s.value).collect()),
+                        counts: Some(samples.iter().map(|s| f64::from(s.rate)).collect()),
+                        timestamp,
+                        dimensions,
+                        ..Default::default()
+                    }),
+                    MetricValue::Set { values } => Some(MetricDatum {
+                        metric_name,
+                        value: Some(values.len() as f64),
+                        timestamp,
+                        dimensions,
+                        ..Default::default()
+                    }),
+                    MetricValue::Gauge { value } => Some(MetricDatum {
+                        metric_name,
+                        value: Some(value),
+                        timestamp,
+                        dimensions,
+                        ..Default::default()
+                    }),
+                    _ => None,
                 }
             })
             .collect()
     }
 }
 
+struct AwsCloudwatchMetricNormalize;
+
+impl MetricNormalize for AwsCloudwatchMetricNormalize {
+    fn apply_state(state: &mut MetricSet, metric: Metric) -> Option<Metric> {
+        match &metric.data.value {
+            MetricValue::Gauge { .. } => state.make_absolute(metric),
+            _ => state.make_incremental(metric),
```

Comment:
Is it semantically correct to just change the metric kind like this? I would expect that the kind limit which metrics we can pass to a given sink, not that we can just adjust the kind to match what the sink expects. If we can just change it then it seems like there's no point in tracking the kind in the first place.

Reply:
There is more to `make_absolute` and `make_incremental` here than just flipping the `kind` bit. The former uses the state information to accumulate the value of incremental metrics, starting from a zero value, to produce an absolute total. The latter uses a previous absolute value to calculate a delta.

Reply:
Ah! I misread, sorry about that.